### ISSUE TITLE
[12.0->13.0][base] error on `base` model when running rename_ir_module_category function related to #2289 issue

### ISSUE THREAD

**Issue by veetoki** [OCA/OpenUpgrade#2303]
Hello,
Thank you very much for the fix on [OCA/OpenUpgrade#2289](https://github.com/OCA/openupgradelib/commit/49827a8).
I have another problem related to `rename_ir_module_category` function.
It is the same problem with `ir_module_category` before but now with `base`.
```
2020-04-13 08:16:04,916 3662 CRITICAL k12_001 odoo.service.server: Failed to initialize database `k12_001`. 
Traceback (most recent call last):
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/service/server.py", line 1179, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/loading.py", line 413, in load_modules
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/loading.py", line 190, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/migration.py", line 185, in migrate_module
    migrate(self.cr, installed_version)
  File "/usr/local/lib/python3.6/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.6/dist-packages/openupgradelib/openupgrade_merge_records.py", line 556, in merge_records
    _change_foreign_key_refs(*args3)
  File "/usr/local/lib/python3.6/dist-packages/openupgradelib/openupgrade_merge_records.py", line 41, in _change_foreign_key_refs
    with env.cr.savepoint():
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/sql_db.py", line 427, in savepoint
    flush_env(self)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/sql_db.py", line 67, in flush_env
    env['base'].flush()
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'base'
```

**Comment by khaeusler** [comment]
Hello

we have also the problem with the base key error
openupgradelib and openupgrade are up2date

```
[…]
2020-04-15 20:23:41,286 1193 DEBUG KH_4 OpenUpgrade: 1 rows affected after 0:00:00.000191 running UPDATE ir_model_data SET module = 'base', name = 'module_category_human_resources_recruitment' WHERE module = 'base' and name = 'module_category_hr_recruitment' 
2020-04-15 20:23:41,348 1193 ERROR KH_4 OpenUpgrade: base: error in migration script base/migrations/13.0.1.3/pre-migration.py: 'base' 
2020-04-15 20:23:41,348 1193 ERROR KH_4 OpenUpgrade: 'base' 
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 556, in merge_records
    _change_foreign_key_refs(*args3)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 41, in _change_foreign_key_refs
    with env.cr.savepoint():
  File "/usr/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 428, in savepoint
    flush_env(self)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 68, in flush_env
    env['base'].flush()
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'base'
2020-04-15 20:23:41,389 1193 WARNING KH_4 odoo.modules.loading: Transient module states were reset 
2020-04-15 20:23:41,414 1193 ERROR KH_4 odoo.modules.registry: Failed to load registry 
Traceback (most recent call last):
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 413, in load_modules
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 190, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/lib/odoo/odoo/modules/migration.py", line 185, in migrate_module
    migrate(self.cr, installed_version)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 556, in merge_records
    _change_foreign_key_refs(*args3)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 41, in _change_foreign_key_refs
    with env.cr.savepoint():
  File "/usr/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 428, in savepoint
    flush_env(self)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 68, in flush_env
    env['base'].flush()
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'base'
2020-04-15 20:23:41,415 1193 CRITICAL KH_4 odoo.service.server: Failed to initialize database `KH_4`. 
Traceback (most recent call last):
  File "/opt/odoo/lib/odoo/odoo/service/server.py", line 1179, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 413, in load_modules
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 190, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/lib/odoo/odoo/modules/migration.py", line 185, in migrate_module
    migrate(self.cr, installed_version)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 556, in merge_records
    _change_foreign_key_refs(*args3)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 41, in _change_foreign_key_refs
    with env.cr.savepoint():
  File "/usr/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 428, in savepoint
    flush_env(self)
  File "/opt/odoo/lib/odoo/odoo/sql_db.py", line 68, in flush_env
    env['base'].flush()
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'base'
```

**Comment by pedrobaeza** [comment]
OK, please stop flooding this with the exact same error. I will check.

**Comment by yucer** [comment]
I don't know if this would be useful, but after a debug one can say that in that point *env* has no model in that point.

Also in the in call to *_change_foreign_key_refs*. So if one adds:

```python
if env:
  _change_foreign_key_refs(*args3)
```

The error doesn't appear. But that may not be the correct solution. Because that call may be needed.

 I am just providing more info to @pedrobaeza

**Comment by pedrobaeza** [comment]
Yes, all the problems come from using `merge_records` on base pre-migration not having yet a registry nor env.

**Comment by pedrobaeza** [comment]
Can you please try with https://github.com/OCA/openupgradelib/pull/204?

For installing this "version" of the library, do:
```
pip install --ignore-installed git+https://github.com/Tecnativa/openupgradelib.git@merge_records-base_pre_migration
```

**Comment by khaeusler** [comment]
Do you  need only the debug output?

```
2020-04-18 12:08:18,909 1274 DEBUG KH_4 OpenUpgrade: 8 rows affected after 0:00:00.013890 running 
        UPDATE ir_translation SET res_id = 127
        WHERE type = 'model' AND res_id in (70)
        AND name like 'ir.module.category' || ',%' 
2020-04-18 12:08:18,910 1274 ERROR KH_4 OpenUpgrade: base: error in migration script base/migrations/13.0.1.3/pre-migration.py: 'ir.module.category' 
2020-04-18 12:08:18,910 1274 ERROR KH_4 OpenUpgrade: 'ir.module.category' 
```

**Comment by pedrobaeza** [comment]
Well, I need the stack trace if any. Or is it working?

**Comment by khaeusler** [comment]
```
2020-04-18 12:13:09,540 249 DEBUG KH_5 OpenUpgrade: 8 rows affected after 0:00:00.007030 running 
        UPDATE ir_translation SET res_id = 127
        WHERE type = 'model' AND res_id in (70)
        AND name like 'ir.module.category' || ',%' 
2020-04-18 12:13:09,540 249 ERROR KH_5 OpenUpgrade: base: error in migration script base/migrations/13.0.1.3/pre-migration.py: 'ir.module.category' 
2020-04-18 12:13:09,540 249 ERROR KH_5 OpenUpgrade: 'ir.module.category' 
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 567, in merge_records
    _delete_records_sql(env, model_name, record_ids, target_record_id)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 477, in _delete_records_sql
    (env[model_name]._table, tuple(record_ids)),
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'ir.module.category'
2020-04-18 12:13:09,575 249 WARNING KH_5 odoo.modules.loading: Transient module states were reset 
2020-04-18 12:13:09,599 249 ERROR KH_5 odoo.modules.registry: Failed to load registry 
Traceback (most recent call last):
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 413, in load_modules
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 190, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/lib/odoo/odoo/modules/migration.py", line 185, in migrate_module
    migrate(self.cr, installed_version)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 567, in merge_records
    _delete_records_sql(env, model_name, record_ids, target_record_id)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 477, in _delete_records_sql
    (env[model_name]._table, tuple(record_ids)),
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'ir.module.category'
2020-04-18 12:13:09,600 249 CRITICAL KH_5 odoo.service.server: Failed to initialize database `KH_5`. 
Traceback (most recent call last):
  File "/opt/odoo/lib/odoo/odoo/service/server.py", line 1179, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 413, in load_modules
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)
  File "/opt/odoo/lib/odoo/odoo/modules/loading.py", line 190, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/lib/odoo/odoo/modules/migration.py", line 185, in migrate_module
    migrate(self.cr, installed_version)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade.py", line 1778, in wrapped_function
    if use_env2 else cr, version)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 750, in migrate
    rename_ir_module_category(env)
  File "/opt/odoo/lib/odoo/odoo/addons/base/migrations/13.0.1.3/pre-migration.py", line 719, in rename_ir_module_category
    method="sql", model_table="ir_module_category")
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 567, in merge_records
    _delete_records_sql(env, model_name, record_ids, target_record_id)
  File "/usr/local/lib/python3.7/dist-packages/openupgradelib/openupgrade_merge_records.py", line 477, in _delete_records_sql
    (env[model_name]._table, tuple(record_ids)),
  File "/opt/odoo/lib/odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "/opt/odoo/lib/odoo/odoo/modules/registry.py", line 178, in __getitem__
    return self.models[model_name]
KeyError: 'ir.module.category'
```

**Comment by khaeusler** [comment]
Any news on this? What can we do to help resolve this issue? Are more tests or traces needed? Or should I create a new issue with this new(?) error?

**Comment by pedrobaeza** [comment]
No, just wait for me that I have time to analyze it or fix it yourself.

**Comment by khaeusler** [comment]
Hello Pedro

https://github.com/khaeusler/openupgradelib/commit/ee6b948e3ee052990a981c428906816c46006d1c

Is it possible that an incorrect argument was passed in the res_model column?
Line 479 and 483: https://github.com/khaeusler/openupgradelib/commit/ee6b948e3ee052990a981c428906816c46006d1c#diff-652fafcb146ddf6218c7b4c9539f20b4R479

The error no longer occurs with this patch, but I get another error with the category that was previously deleted.
```
DEBUG KH_4 OpenUpgrade: 26 rows affected after 0:00:00.014130 running UPDATE ir_module_module
SET "category_id" = 168
WHERE "category_id" in (18)

DEBUG KH_4 OpenUpgrade: 8 rows affected after 0:00:00.008154 running
UPDATE ir_translation SET res_id = 168
WHERE type = 'model' AND res_id in (18)
AND name like 'ir.module.category' || ',%'
DEBUG KH_4 OpenUpgrade: 1 rows affected after 0:00:00.003633 running DELETE FROM ir_model_data WHERE model = 'ir.module.category' AND id IN (18)
DEBUG KH_4 OpenUpgrade: 0 rows affected after 0:00:00.000544 running DELETE FROM ir_attachment WHERE res_model = 'ir.module.category' AND id IN (18)
DEBUG KH_4 OpenUpgrade: 1 rows affected after 0:00:00.001144 running DELETE FROM ir_module_category WHERE id IN (18)
[…]
ERROR KH_4 odoo.sql_db: bad query: UPDATE "ir_module_module" SET "write_uid"=1,"website"='https://www.cybrosys.com',"license"='AGPL-3',"category_id"=18,"write_date"=(now() at time zone 'UTC') WHERE id IN (3934)
ERROR: insert or update on table "ir_module_module" violates foreign key constraint "ir_module_module_category_id_fkey"
DETAIL:  Key (category_id)=(18) is not present in table "ir_module_category".
```
Why is the category_id set from 168 back to 18, before it was set from 18 to 168.
The problematic addons come from https://github.com/CybroOdoo/CybroAddons/tree/13.0, e.g. pos_mrp_order. When I delete all these Addons from this branch, no error occurs.

Greetings

**Comment by kos94ok-3D** [comment]
@khaeusler You use model_name instead of model_table
Please check lines 479 and 483 in khaeusler/openupgradelib@ee6b948#diff-652fafcb146ddf6218c7b4c9539f20b4R479

**Comment by pedrobaeza** [comment]
Fixed in #204 including @khaeusler's patch, and yes, the deletion was incorrect previously, leaving some garbage, so thanks also for that bugfix.

**Comment by khaeusler** [comment]
Thanks, the base key error is fixed
I've created a new issue #2323; the updating of the modul category is still available.
